From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Fri, 7 Jul 2023 19:14:46 +0800
Subject: [PATCH] KeYi Player skull api


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 14c92390aeaeee518f43778ccd5fc820fdc62689..2b75b3db93a58cca79b493de4eb628e29a3a95c2 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -47,6 +47,8 @@ import org.bukkit.scoreboard.Scoreboard;
 import org.jetbrains.annotations.ApiStatus;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
 
 /**
  * Represents a player, connected or not
@@ -3241,4 +3243,22 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
     @Override
     Spigot spigot();
     // Spigot end
+
+    // KeYi start
+    /**
+     * Get a skull item of a player.
+     * This method runs on main thread, which may freeze the server.
+     *
+     * @return A ItemStack of the skull of the player
+     */
+    ItemStack getSkull();
+
+    /**
+     * Get a skull item of a player.
+     * This method runs on main thread, which may freeze the server.
+     *
+     * @return A CompletableFuture the of ItemStack of the skull of the player
+     */
+    CompletableFuture<ItemStack> getSkullAsynchronously();
+    // KeYi end
 }
