From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Fri, 25 Aug 2023 22:23:40 +0800
Subject: [PATCH] Leaves fakeplayer support


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 26b32a5935aaed9b1fe356af0208a778144aac87..166e67c1d61d2e8e409051d11da3707f953c8191 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -57,6 +57,7 @@ import org.jetbrains.annotations.Contract;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 import io.papermc.paper.util.JarManifests; // Paper
+import top.leavesmc.leaves.entity.BotManager;
 
 /**
  * Represents the Bukkit core, for version and Server singleton handling
@@ -2760,6 +2761,17 @@ public final class Bukkit {
     }
     // Folia end - region threading API
 
+    // Leaves start - Bot API
+    /**
+     * Returns a bot manager.
+     *
+     * @return Bot Manager
+     */
+    public static @NotNull BotManager getBotManager() {
+        return server.getBotManager();
+    }
+    // Leaves end - Bot API
+
     @NotNull
     public static Server.Spigot spigot() {
         return server.spigot();
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index b6a8217b97842cd07af24fadb21e0e44226e5bb9..c06409071463d79c3e2484dc68b47b1ef24a581d 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -57,12 +57,22 @@ import org.bukkit.util.CachedServerIcon;
 import org.jetbrains.annotations.Contract;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
+import top.leavesmc.leaves.entity.BotManager;
 
 /**
  * Represents a server implementation.
  */
 public interface Server extends PluginMessageRecipient, net.kyori.adventure.audience.ForwardingAudience { // Paper
 
+    // Leaves start - Bot API
+    /**
+     * Returns a bot manager.
+     *
+     * @return Bot Manager
+     */
+    @NotNull BotManager getBotManager();
+    // Leaves end - Bot API
+
     /**
      * Returns the de facto plugins directory, generally used for storing plugin jars to be loaded,
      * as well as their {@link org.bukkit.plugin.Plugin#getDataFolder() data folders}.
