From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Wed, 9 Aug 2023 09:33:59 +0800
Subject: [PATCH] Make fakeplayer join or left message more like normal player


diff --git a/src/main/java/top/leavesmc/leaves/event/bot/BotCreateEvent.java b/src/main/java/top/leavesmc/leaves/event/bot/BotCreateEvent.java
index 7cf1eb4eb3d2fe9310f9272ec53208632b87b49b..0a64f182a171518732bb410dad11fc84965b9c23 100644
--- a/src/main/java/top/leavesmc/leaves/event/bot/BotCreateEvent.java
+++ b/src/main/java/top/leavesmc/leaves/event/bot/BotCreateEvent.java
@@ -18,6 +18,7 @@ public class BotCreateEvent extends Event implements Cancellable {
     private String joinMessage;
     private Location createLocation;
     private boolean cancel = false;
+    private boolean joinMessageModified = false; //Molia - Make fake player more like normal player if the join message is not modified
 
     public BotCreateEvent(@NotNull final String who, @NotNull final String skin, @NotNull final Location createLocation, @Nullable final String joinMessage) {
         this.bot = who;
@@ -51,6 +52,7 @@ public class BotCreateEvent extends Event implements Cancellable {
      * @param joinMessage join message. If null, no message will be sent
      */
     public void setJoinMessage(@Nullable String joinMessage) {
+        this.joinMessageModified = true;
         this.joinMessage = joinMessage;
     }
 
@@ -93,6 +95,12 @@ public class BotCreateEvent extends Event implements Cancellable {
         this.cancel = cancel;
     }
 
+    //Molia start - Make fake player more like normal player if the join message is not modified
+    public boolean isJoinMessageModified(){
+        return this.joinMessageModified;
+    }
+    //Molia end
+
     @Override
     @NotNull
     public HandlerList getHandlers() {
