From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Fri, 25 Aug 2023 20:48:28 +0800
Subject: [PATCH] Fix build


diff --git a/build.gradle.kts b/build.gradle.kts
index 9ee90bc753e65f9dd532f28ec75cc16bf16486c8..f42dbf797dddc14324ee720c2aefc0310f6ac82e 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -14,7 +14,7 @@ val alsoShade: Configuration by configurations.creating
 
 dependencies {
     // Folia start
-    implementation(project(":folia-api"))
+    implementation(project(":molia-api")) //Molia
     implementation("io.papermc.paper:paper-mojangapi:1.19.3-R0.1-SNAPSHOT") {
         exclude("io.papermc.paper", "paper-api")
     }
@@ -76,7 +76,7 @@ tasks.jar {
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
             "Implementation-Title" to "CraftBukkit",
-            "Implementation-Version" to "git-Folia-$implementationVersion", // Folia
+            "Implementation-Version" to "git-Molia-$implementationVersion", // Folia //Molia
             "Implementation-Vendor" to date, // Paper
             "Specification-Title" to "Bukkit",
             "Specification-Version" to project.version,
