From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Fri, 31 Mar 2023 19:09:11 +0800
Subject: [PATCH] Molia build setting changes


diff --git a/build.gradle.kts b/build.gradle.kts
index dd53b1ffaff8dbaea0cf21c489a90162a496a026..fa27bf380c9913942caabff1e11310ad8b287769 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -8,7 +8,7 @@ plugins {
 
 dependencies {
     // Folia start
-    implementation(project(":folia-api"))
+    implementation(project(":molia-api")) //Molia
     implementation("io.papermc.paper:paper-mojangapi:1.19.3-R0.1-SNAPSHOT") {
         exclude("io.papermc.paper", "paper-api")
     }
@@ -66,7 +66,7 @@ tasks.jar {
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
             "Implementation-Title" to "CraftBukkit",
-            "Implementation-Version" to "git-Folia-$implementationVersion", // Folia
+            "Implementation-Version" to "git-Molia-$implementationVersion", // Folia //Molia
             "Implementation-Vendor" to date, // Paper
             "Specification-Title" to "Bukkit",
             "Specification-Version" to project.version,
