From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Sat, 12 Aug 2023 09:22:08 +0800
Subject: [PATCH] Rework doubling chunk sending fix


diff --git a/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java b/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java
index 9990fccc741a139c57528f70157ae63dbe3fd8e8..f3a54101728f1f896731dc545d06e69fcd27051d 100644
--- a/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java
+++ b/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java
@@ -30,6 +30,7 @@ import net.minecraft.world.level.chunk.ChunkStatus;
 import net.minecraft.world.level.chunk.LevelChunk;
 import net.minecraft.world.level.levelgen.BelowZeroRetrogen;
 import org.apache.commons.lang3.mutable.MutableObject;
+import org.apache.logging.log4j.LogManager;
 import org.bukkit.craftbukkit.entity.CraftPlayer;
 import org.bukkit.entity.Player;
 import java.lang.invoke.VarHandle;
@@ -480,17 +481,24 @@ public class RegionizedPlayerChunkLoader {
         }
 
         private void sendChunk(final int chunkX, final int chunkZ) {
+            //Molia start - Rework it
+            if (this.sentChunks.contains(CoordinateUtils.getChunkKey(chunkX,chunkZ))){
+                LogManager.getLogger().warn("There was already sent a chunk!Skipping sending");
+                return;
+            }
+
             if (this.sentChunks.add(CoordinateUtils.getChunkKey(chunkX, chunkZ))) {
                 this.world.getChunkSource().chunkMap.updateChunkTracking(this.player,
                     new ChunkPos(chunkX, chunkZ), new MutableObject<>(), false, true); // unloaded, loaded
-                return;
             }
-            // Kaiiju - Kick player instead of crashing
+
+            /*// Kaiiju - Kick player instead of crashing
             String errorMsg = "Already sent chunk [" + chunkX + ", " + chunkZ + "] in world " + this.world;
             this.player.getBukkitEntity().kickPlayer(errorMsg);
             org.bukkit.Bukkit.getLogger().severe(errorMsg);
             // throw new IllegalStateException();
-            // Kaiiju end
+            // Kaiiju end*/
+            //Molia end
         }
 
         private void sendUnloadChunk(final int chunkX, final int chunkZ) {
