From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Fri, 25 Aug 2023 21:02:58 +0800
Subject: [PATCH] Fix can not move entity off main


diff --git a/src/main/java/me/earthme/molia/MoliaConfig.java b/src/main/java/me/earthme/molia/MoliaConfig.java
index d334b3aa4f311f57f2c41ced6c2265c817b78a76..92ad4c7105d5a5d8040d52c85d2274abf247d226 100644
--- a/src/main/java/me/earthme/molia/MoliaConfig.java
+++ b/src/main/java/me/earthme/molia/MoliaConfig.java
@@ -8,12 +8,14 @@ import java.io.IOException;
 
 public class MoliaConfig {
     //All values
+    public static boolean fixCanNotMoveEntityOffMain = true;
     //
 
     private static final YamlConfiguration configEntry = new YamlConfiguration();
     private static final File CONFIG_FILE = new File("molia.yml");
 
     public static void loadAllValues(){
+        fixCanNotMoveEntityOffMain = getBoolean("fixes.fix_can_not_move_entity_off_main",fixCanNotMoveEntityOffMain);
         save();
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index fa35d2c1c8de225acd68e08f15976c92f7ab82aa..6110fef8afe827bb8b724853973f41e8a32b28a1 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -7,6 +7,7 @@ import com.google.common.collect.Lists;
 import com.google.common.collect.Sets;
 import com.google.common.collect.UnmodifiableIterator;
 import com.mojang.logging.LogUtils;
+import io.papermc.paper.util.TickThread;
 import it.unimi.dsi.fastutil.objects.Object2DoubleArrayMap;
 import it.unimi.dsi.fastutil.objects.Object2DoubleMap;
 import java.util.Arrays;
@@ -24,6 +25,8 @@ import java.util.function.BiConsumer;
 import java.util.function.Predicate;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
+
+import me.earthme.molia.MoliaConfig;
 import net.minecraft.BlockUtil;
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
@@ -767,6 +770,22 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     }
 
     public void setPos(double x, double y, double z) {
+        //Molia start - Teleport if the pos is in another region
+        if (TickThread.isTickThread()){ //Skip for chunk system worker
+            if (MoliaConfig.fixCanNotMoveEntityOffMain && !TickThread.isTickThreadFor((ServerLevel) this.level,(int)x >> 4,(int)z >> 4)){
+                this.teleportAsync(
+                        ((ServerLevel) this.level),
+                        new Vec3(x,y,z),
+                        this.getYRot(),
+                        this.getXRot(),
+                        null,
+                        PlayerTeleportEvent.TeleportCause.UNKNOWN,
+                        Entity.TELEPORT_FLAG_LOAD_CHUNK | Entity.TELEPORT_FLAG_TELEPORT_PASSENGERS, null
+                );
+                return;
+            }
+        }
+        //Molia end
         this.setPosRaw(x, y, z, true); // Paper - force bounding box update
         // this.setBoundingBox(this.makeBoundingBox()); // Paper - move into setPositionRaw
     }
