From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: M2ke4U <wangxyper@163.com>
Date: Sat, 17 Jun 2023 08:27:07 +0800
Subject: [PATCH] Reduce sensor work


diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index a8b23b1594d2b39568c68c93a8a1b936457672bc..b25021a85f8c9cb9df849c56e2ee825be3d6106b 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -918,10 +918,16 @@ public abstract class Mob extends LivingEntity implements Targeting {
             return;
         }
         // Paper end
-        this.level().getProfiler().push("sensing");
-        this.sensing.tick();
-        this.level().getProfiler().pop();
-        int i = this.tickCount + this.getId(); // Folia - region threading
+        int i = this.tickCount + this.getId(); // Folia - region threading //Molia move up
+
+        //Molia start - reduce sensor work
+        if (i % 10 == 0){
+            this.level().getProfiler().push("sensing");
+            this.sensing.tick();
+            this.level().getProfiler().pop();
+        }
+        //Molia end
+
 
         if (i % 2 != 0 && this.tickCount > 1) {
             this.level().getProfiler().push("targetSelector");
