From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Thu, 24 Aug 2023 14:12:35 +0800
Subject: [PATCH] Init config on server startup


diff --git a/src/main/java/me/earthme/molia/MoliaConfig.java b/src/main/java/me/earthme/molia/MoliaConfig.java
index 0da9fbb069b1de552857a303ffead40eefc5c8fd..e45d2416ee7ca34d885763a51d795c67cac2b10b 100644
--- a/src/main/java/me/earthme/molia/MoliaConfig.java
+++ b/src/main/java/me/earthme/molia/MoliaConfig.java
@@ -2,14 +2,25 @@ package me.earthme.molia;
 
 import org.bukkit.configuration.InvalidConfigurationException;
 import org.bukkit.configuration.file.YamlConfiguration;
-
+import com.google.common.collect.Lists;
 import java.io.File;
 import java.io.IOException;
+import java.util.List;
 
 public class MoliaConfig {
     //All values
+
+    //Optimizations
     public static double entityWakeUpDurationRatioStandardDeviation = 0.2;
     public static boolean enableSuffocationOptimization = true;
+    //Gameplay
+    public static boolean fakeplayerSupport = false;
+    public static boolean alwaysSendFakeplayerData = true;
+    public static boolean fakeplayerSkipSleep = true;
+    public static boolean openFakeplayerInventory = true;
+    public static boolean fakeplayerResident = false;
+    public static List<String> unableFakeplayerNames = Lists.newArrayList();
+    public static int fakeplayerLimit = 10;
     //
 
     private static final YamlConfiguration configEntry = new YamlConfiguration();
@@ -18,6 +29,11 @@ public class MoliaConfig {
     public static void loadOrInitAll(){
         entityWakeUpDurationRatioStandardDeviation = getDouble("optimizations.entity_wake_up_duration_ratio_standard_deviation",entityWakeUpDurationRatioStandardDeviation);
         enableSuffocationOptimization = getBoolean("optimizations.enable_suffocation_optimization",enableSuffocationOptimization);
+
+        //Gameplay
+        fakeplayerSupport = getBoolean("gameplay.leaves.fakeplayer.enabled",fakeplayerSupport);
+        fakeplayerLimit = getInt("gameplay.leaves.fakeplayer.fake_player_limit",fakeplayerLimit);
+        unableFakeplayerNames = getStringList("gameplay.leaves.fakeplayer.name_black_list",unableFakeplayerNames);
     }
 
     public static void init(){
@@ -28,6 +44,8 @@ public class MoliaConfig {
             e.printStackTrace();
         }
         configEntry.options().copyDefaults(true);
+
+        loadOrInitAll();
     }
 
     public static void save(){
@@ -62,4 +80,9 @@ public class MoliaConfig {
         configEntry.addDefault(key,def);
         return configEntry.getDouble(key);
     }
+
+    public static List<String> getStringList(String key,List<String> def){
+        configEntry.addDefault(key,def);
+        return configEntry.getStringList(key);
+    }
 }
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 24a2d3f496727790f63cb66a2534d4423555fa3c..95f3c3ee503aa3c8db58feb54ec4aadda65766af 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -19,6 +19,8 @@ import java.util.Locale;
 import java.util.Optional;
 import java.util.function.BooleanSupplier;
 import javax.annotation.Nullable;
+
+import me.earthme.molia.MoliaConfig;
 import net.minecraft.DefaultUncaughtExceptionHandler;
 import net.minecraft.DefaultUncaughtExceptionHandlerWithName;
 import net.minecraft.SharedConstants;
@@ -199,6 +201,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
             this.setPreventProxyConnections(dedicatedserverproperties.preventProxyConnections);
             this.setLocalIp(dedicatedserverproperties.serverIp);
         }
+        MoliaConfig.init(); //Molia
         // Spigot start
         this.setPlayerList(new DedicatedPlayerList(this, this.registries(), this.playerDataStorage));
         org.spigotmc.SpigotConfig.init((java.io.File) options.valueOf("spigot-settings"));
