From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Thu, 24 Aug 2023 14:20:43 +0800
Subject: [PATCH] Save config on loaded


diff --git a/src/main/java/me/earthme/molia/MoliaConfig.java b/src/main/java/me/earthme/molia/MoliaConfig.java
index a6d8e9e6b772ffc1f67eac5108e5a3f97ff3609d..10e9fd37b41f2acf2db60b8a532d68175cd06a1e 100644
--- a/src/main/java/me/earthme/molia/MoliaConfig.java
+++ b/src/main/java/me/earthme/molia/MoliaConfig.java
@@ -63,6 +63,7 @@ public class MoliaConfig {
         configEntry.options().copyDefaults(true);
 
         loadOrInitAll();
+        save();
     }
 
     public static void save(){
