From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: M2ke4U <wangxyper@163.com>
Date: Sat, 17 Jun 2023 08:29:54 +0800
Subject: [PATCH] Remove sand duping fixes


diff --git a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
index 34b3541603b8cca16c7d62f3981d7ce3e8be0dbe..146254dc045e529a0fc4622a91c7b6d022a6b42d 100644
--- a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -132,11 +132,13 @@ public class FallingBlockEntity extends Entity {
 
     @Override
     public void tick() {
-        // Paper start - fix sand duping
+        //Molia start - we don't need this
+        /*// Paper start - fix sand duping
         if (this.isRemoved()) {
             return;
         }
-        // Paper end - fix sand duping
+        // Paper end - fix sand duping*/
+        //Molia end
         if (this.blockState.isAir()) {
             this.discard();
         } else {
@@ -149,11 +151,13 @@ public class FallingBlockEntity extends Entity {
 
             this.move(MoverType.SELF, this.getDeltaMovement());
 
-            // Paper start - fix sand duping
+            //Molia start - we don't need this
+            /*// Paper start - fix sand duping
             if (this.isRemoved()) {
                 return;
             }
-            // Paper end - fix sand duping
+            // Paper end - fix sand duping*/
+            //Molia end
 
             // Paper start - Configurable EntityFallingBlock height nerf
             if (this.level().paperConfig().fixes.fallingBlockHeightNerf.test(v -> this.getY() > v)) {
