From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: M2ke4U <wangxyper@163.com>
Date: Sun, 18 Jun 2023 08:19:50 +0800
Subject: [PATCH] Add config for folia plugin checking


diff --git a/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginProviderFactory.java b/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginProviderFactory.java
index dc81e1a9faa54ba677bede6d7ca2414b1d7b6003..198420d62bf983df465e65df518a38b73f09b114 100644
--- a/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginProviderFactory.java
+++ b/src/main/java/io/papermc/paper/plugin/provider/type/paper/PaperPluginProviderFactory.java
@@ -18,6 +18,8 @@ import java.nio.file.Path;
 import java.util.jar.JarEntry;
 import java.util.jar.JarFile;
 import java.util.logging.Logger;
+
+import me.earthme.molia.MoliaConfig;
 import net.kyori.adventure.text.logger.slf4j.ComponentLogger;
 
 class PaperPluginProviderFactory implements PluginTypeFactory<PaperPluginParent, PaperPluginMeta> {
@@ -25,7 +27,7 @@ class PaperPluginProviderFactory implements PluginTypeFactory<PaperPluginParent,
     @Override
     public PaperPluginParent build(JarFile file, PaperPluginMeta configuration, Path source) throws Exception {
         // Folia start - block plugins not marked as supported
-        if (!configuration.isFoliaSupported()) {
+        if (!configuration.isFoliaSupported() && !MoliaConfig.disableFoliaPluginCheck) { //Molia - add config for folia plugin check
             throw new RuntimeException("Could not load plugin '" + configuration.getDisplayName() + "' as it is not marked as supporting Folia!");
         }
         // Folia end - block plugins not marked as supported
diff --git a/src/main/java/io/papermc/paper/plugin/provider/type/spigot/SpigotPluginProviderFactory.java b/src/main/java/io/papermc/paper/plugin/provider/type/spigot/SpigotPluginProviderFactory.java
index ad13d60eeb94a75c97b3d0696c39d834d6082bfa..25276432975347fc3a619a1c3d1f52d4a4c58704 100644
--- a/src/main/java/io/papermc/paper/plugin/provider/type/spigot/SpigotPluginProviderFactory.java
+++ b/src/main/java/io/papermc/paper/plugin/provider/type/spigot/SpigotPluginProviderFactory.java
@@ -2,6 +2,7 @@ package io.papermc.paper.plugin.provider.type.spigot;
 
 import io.papermc.paper.plugin.provider.configuration.serializer.constraints.PluginConfigConstraints;
 import io.papermc.paper.plugin.provider.type.PluginTypeFactory;
+import me.earthme.molia.MoliaConfig;
 import org.bukkit.plugin.InvalidDescriptionException;
 import org.bukkit.plugin.PluginDescriptionFile;
 import org.yaml.snakeyaml.error.YAMLException;
@@ -18,7 +19,7 @@ class SpigotPluginProviderFactory implements PluginTypeFactory<SpigotPluginProvi
     @Override
     public SpigotPluginProvider build(JarFile file, PluginDescriptionFile configuration, Path source) throws Exception {
         // Folia start - block plugins not marked as supported
-        if (!configuration.isFoliaSupported()) {
+        if (!configuration.isFoliaSupported() && !MoliaConfig.disableFoliaPluginCheck) { //Molia - add config for folia plugin check
             throw new RuntimeException("Could not load plugin '" + configuration.getDisplayName() + "' as it is not marked as supporting Folia!");
         }
         // Folia end - block plugins not marked as supported
diff --git a/src/main/java/me/earthme/molia/MoliaConfig.java b/src/main/java/me/earthme/molia/MoliaConfig.java
index 63f54137a3cdf05a39d699364704ca942fcec042..d0df1c3ccefbe845b67ccb870d3959d4df3ec1fb 100644
--- a/src/main/java/me/earthme/molia/MoliaConfig.java
+++ b/src/main/java/me/earthme/molia/MoliaConfig.java
@@ -14,6 +14,7 @@ public class MoliaConfig {
     public static double entityWakeUpDurationRatioStandardDeviation = 0.2;
     public static int acquirePoiForStuckEntity = 200;
     public static String serverModName = "Molia";
+    public static boolean disableFoliaPluginCheck = false;
 
     private static final YamlConfiguration configEntry = new YamlConfiguration();
     private static final File CONFIG_FILE = new File("molia.yml");
@@ -33,6 +34,7 @@ public class MoliaConfig {
 
         serverModName = getString("misc.server_mod_name","Molia");
 
+        disableFoliaPluginCheck = getBoolean("unsafe.disable_folia_plugin_check",disableFoliaPluginCheck);
         save();
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
index 61e40979a68017eb730f14a1a0978402f44c1865..9c15c406e1b5e3b89cf77f33ccc9a62571b69139 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
@@ -23,6 +23,8 @@ import java.util.Map;
 import java.util.Map.Entry;
 import java.util.logging.Level;
 import java.util.logging.Logger;
+
+import me.earthme.molia.MoliaConfig;
 import net.minecraft.SharedConstants;
 import net.minecraft.advancements.critereon.DeserializationContext;
 import net.minecraft.core.registries.BuiltInRegistries;
@@ -374,7 +376,7 @@ public final class CraftMagicNumbers implements UnsafeValues {
         int minimumIndex = CraftMagicNumbers.SUPPORTED_API.indexOf(minimumVersion);
 
         // Folia start - block plugins not marked as supported
-        if (!pdf.isFoliaSupported()) {
+        if (!pdf.isFoliaSupported() && !MoliaConfig.disableFoliaPluginCheck) { //Molia - add config for folia plugin check
             throw new InvalidPluginException("Plugin " + pdf.getFullName() + " is not marked as supporting regionised multithreading");
         }
         // Folia end - block plugins not marked as supported
