From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Thu, 24 Aug 2023 18:55:16 +0800
Subject: [PATCH] Gale Reduce villager item re-pickup


diff --git a/src/main/java/me/earthme/molia/MoliaConfig.java b/src/main/java/me/earthme/molia/MoliaConfig.java
index 10e9fd37b41f2acf2db60b8a532d68175cd06a1e..14e5b716ac052228cbeb4ae0d2452cf1f025ba34 100644
--- a/src/main/java/me/earthme/molia/MoliaConfig.java
+++ b/src/main/java/me/earthme/molia/MoliaConfig.java
@@ -17,6 +17,7 @@ public class MoliaConfig {
     //Optimizations
     public static double entityWakeUpDurationRatioStandardDeviation = 0.2;
     public static boolean enableSuffocationOptimization = true;
+    public static int villagerRepickItemDelay = 100;
     //Gameplay
     public static boolean fakeplayerSupport = false;
     public static boolean alwaysSendFakeplayerData = true;
@@ -42,6 +43,7 @@ public class MoliaConfig {
     public static void loadOrInitAll(){
         entityWakeUpDurationRatioStandardDeviation = getDouble("optimizations.entity_wake_up_duration_ratio_standard_deviation",entityWakeUpDurationRatioStandardDeviation);
         enableSuffocationOptimization = getBoolean("optimizations.enable_suffocation_optimization",enableSuffocationOptimization);
+        villagerRepickItemDelay = getInt("optimizations.villager_repick_item_delay",villagerRepickItemDelay);
 
         //Gameplay
         fakeplayerSupport = getBoolean("gameplay.leaves.fakeplayer.enabled",fakeplayerSupport);
diff --git a/src/main/java/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java b/src/main/java/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
index 92bd58010e8c89e361e28aec59447349edbbc028..caabe9a78f613df591270c558fd2ec87a5bdd51a 100644
--- a/src/main/java/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
+++ b/src/main/java/net/minecraft/world/entity/ai/behavior/BehaviorUtils.java
@@ -7,6 +7,8 @@ import java.util.UUID;
 import java.util.function.Predicate;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
+
+import me.earthme.molia.MoliaConfig;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.SectionPos;
 import net.minecraft.server.level.ServerLevel;
@@ -21,6 +23,7 @@ import net.minecraft.world.entity.ai.memory.NearestVisibleLivingEntities;
 import net.minecraft.world.entity.ai.memory.WalkTarget;
 import net.minecraft.world.entity.ai.util.DefaultRandomPos;
 import net.minecraft.world.entity.item.ItemEntity;
+import net.minecraft.world.entity.npc.Villager;
 import net.minecraft.world.item.Item;
 import net.minecraft.world.item.ItemStack;
 import net.minecraft.world.item.ProjectileWeaponItem;
@@ -101,7 +104,16 @@ public class BehaviorUtils {
 
         vec3d2 = vec3d2.normalize().multiply(velocityFactor.x, velocityFactor.y, velocityFactor.z);
         entityitem.setDeltaMovement(vec3d2);
-        entityitem.setDefaultPickUpDelay();
+        // Gale start - EMC - reduce villager item re-pickup
+        if (entity instanceof Villager) {
+            int repickupDelay = MoliaConfig.villagerRepickItemDelay;
+            if (repickupDelay <= -1) {
+                entityitem.setDefaultPickUpDelay();
+            } else {
+                entityitem.pickupDelay = repickupDelay;
+            }
+        }
+        // Gale end - EMC - reduce villager item re-pickup
         // CraftBukkit start
         org.bukkit.event.entity.EntityDropItemEvent event = new org.bukkit.event.entity.EntityDropItemEvent(entity.getBukkitEntity(), (org.bukkit.entity.Item) entityitem.getBukkitEntity());
         entityitem.level().getCraftServer().getPluginManager().callEvent(event);
