From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Sat, 12 Aug 2023 13:52:29 +0800
Subject: [PATCH] Save config on loaded


diff --git a/src/main/java/me/earthme/molia/MoliaConfig.java b/src/main/java/me/earthme/molia/MoliaConfig.java
index a02a687e7e01e6d6cc8d6bb7fe5b293b03018433..31b0a7a3ed69e1b78a782ddef88b7b78ba0e26cb 100644
--- a/src/main/java/me/earthme/molia/MoliaConfig.java
+++ b/src/main/java/me/earthme/molia/MoliaConfig.java
@@ -67,6 +67,8 @@ public class MoliaConfig {
         fakeplayerSupport = getBoolean("gameplay.leaves.fakeplayer.enabled",fakeplayerSupport);
         fakeplayerLimit = getInt("gameplay.leaves.fakeplayer.fake_player_limit",fakeplayerLimit);
         unableFakeplayerNames = getStringList("gameplay.leaves.fakeplayer.name_black_list",unableFakeplayerNames);
+
+        save();
     }
 
     public static void init(){
