From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Fri, 18 Aug 2023 21:09:19 +0800
Subject: [PATCH] Fix fakeplayer issue


diff --git a/src/main/java/me/earthme/molia/MoliaConfig.java b/src/main/java/me/earthme/molia/MoliaConfig.java
index 31b0a7a3ed69e1b78a782ddef88b7b78ba0e26cb..3061d68a2901a559d609d286c3e13710a859a8dd 100644
--- a/src/main/java/me/earthme/molia/MoliaConfig.java
+++ b/src/main/java/me/earthme/molia/MoliaConfig.java
@@ -67,6 +67,13 @@ public class MoliaConfig {
         fakeplayerSupport = getBoolean("gameplay.leaves.fakeplayer.enabled",fakeplayerSupport);
         fakeplayerLimit = getInt("gameplay.leaves.fakeplayer.fake_player_limit",fakeplayerLimit);
         unableFakeplayerNames = getStringList("gameplay.leaves.fakeplayer.name_black_list",unableFakeplayerNames);
+        alwaysSendFakeplayerData = getBoolean("gameplay.leaves.fakeplayer.always_send_data",alwaysSendFakeplayerData);
+        fakeplayerSkipSleep = getBoolean("gameplay.leaves.fakeplayer.fakeplayer_skip_sleep",fakeplayerResident);
+        openFakeplayerInventory = getBoolean("gameplay.leaves.fakeplayer.open_inventory",openFakeplayerInventory);
+        fakeplayerResident = getBoolean("gameplay.leaves.fakeplayer.fakeplayer_resident",fakeplayerResident);
+        if (fakeplayerSupport){
+            initFakePlayer();
+        }
 
         save();
     }
