From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: M2ke4U <wangxyper@163.com>
Date: Thu, 22 Jun 2023 10:33:29 +0800
Subject: [PATCH] Replace something in KeYi's player skull api


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 32209188489beb23ad932c4c735f53180a70f4a8..26d51bb88696e00a7fd194d45c23a8f4b034464f 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -149,6 +149,7 @@ import org.bukkit.craftbukkit.inventory.CraftItemStack;
 import org.bukkit.craftbukkit.map.CraftMapView;
 import org.bukkit.craftbukkit.map.RenderData;
 import org.bukkit.craftbukkit.profile.CraftPlayerProfile;
+import org.bukkit.craftbukkit.scheduler.CraftScheduler;
 import org.bukkit.craftbukkit.scoreboard.CraftScoreboard;
 import org.bukkit.craftbukkit.util.CraftChatMessage;
 import org.bukkit.craftbukkit.util.CraftLocation;
@@ -3244,12 +3245,17 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
 
     @Override
     public CompletableFuture<ItemStack> getSkullAsynchronously() {
+        //Molia start - A better replacement of this code block
+        /*
         ExecutorService executorService = Executors.newCachedThreadPool();
 
         CompletableFuture<ItemStack> future = (CompletableFuture<ItemStack>) executorService.submit(() -> getSkull());
         executorService.shutdown();
 
-        return future;
+        return future;*/
+
+        return CompletableFuture.supplyAsync(this::getSkull,r -> Bukkit.getAsyncScheduler().runNow(CraftScheduler.MINECRAFT, a->r.run()));
+        //Molia end
     }
     // KeYi end
 }
