From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Thu, 24 Aug 2023 18:58:23 +0800
Subject: [PATCH] Gale Cache BlockStatePairKey hash


diff --git a/src/main/java/net/minecraft/world/level/block/Block.java b/src/main/java/net/minecraft/world/level/block/Block.java
index 52dda8e8074e0f894d9ef95b46bd937dae629f14..b1af36578de1c89ef5c67d983c784a800735e5c0 100644
--- a/src/main/java/net/minecraft/world/level/block/Block.java
+++ b/src/main/java/net/minecraft/world/level/block/Block.java
@@ -590,11 +590,18 @@ public class Block extends BlockBehaviour implements ItemLike {
         private final BlockState first;
         private final BlockState second;
         private final Direction direction;
+        private final int hash; // Gale - Lithium - cache BlockStatePairKey hash
 
         public BlockStatePairKey(BlockState self, BlockState other, Direction facing) {
             this.first = self;
             this.second = other;
             this.direction = facing;
+            // Gale start - Lithium - cache BlockStatePairKey hash
+            int hash = this.first.hashCode();
+            hash = 31 * hash + this.second.hashCode();
+            hash = 31 * hash + this.direction.hashCode();
+            this.hash = hash;
+            // Gale end - Lithium - cache BlockStatePairKey hash
         }
 
         public boolean equals(Object object) {
@@ -610,11 +617,7 @@ public class Block extends BlockBehaviour implements ItemLike {
         }
 
         public int hashCode() {
-            int i = this.first.hashCode();
-
-            i = 31 * i + this.second.hashCode();
-            i = 31 * i + this.direction.hashCode();
-            return i;
+            return this.hash; // Gale - Lithium - cache BlockStatePairKey hash
         }
     }
 }
