From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Fri, 25 Aug 2023 22:41:34 +0800
Subject: [PATCH] Add config for trip door feature


diff --git a/src/main/java/me/earthme/molia/MoliaConfig.java b/src/main/java/me/earthme/molia/MoliaConfig.java
index 69dd5cc40ded81c6cb89e1d6c2f042e5369d6e0f..0b83f7265fe63e4e603cf23b356d2e05bec7b733 100644
--- a/src/main/java/me/earthme/molia/MoliaConfig.java
+++ b/src/main/java/me/earthme/molia/MoliaConfig.java
@@ -16,6 +16,7 @@ public class MoliaConfig {
     public static boolean fixDoublingChunkSend = true;
     public static boolean allowGravityBlockDupe = false;
     public static boolean updateTripWireIfRemoving = false;
+    public static boolean breakRedstoneOnTripDoorEarly = true;
 
     public static double entityWakeUpDurationRatioStandardDeviation = 0.2;
     public static boolean enableSuffocationOptimization = true;
@@ -46,6 +47,7 @@ public class MoliaConfig {
         fixDoublingChunkSend = getBoolean("fixes.fix_doubling_chunk_send",fixDoublingChunkSend);
         allowGravityBlockDupe = getBoolean("fixes.allow_gravity_block_dupe",allowGravityBlockDupe);
         updateTripWireIfRemoving = getBoolean("fixes.update_tripwire_if_removing",updateTripWireIfRemoving);
+        breakRedstoneOnTripDoorEarly = getBoolean("fixes.break_redstone_early_on_tripdoor",breakRedstoneOnTripDoorEarly);
 
         entityWakeUpDurationRatioStandardDeviation = getDouble("optimizations.entity_wakeup_duration_ratio_standard_deviation",entityWakeUpDurationRatioStandardDeviation);
         enableSuffocationOptimization = getBoolean("optimizations.enable_suffocation_optimization",enableSuffocationOptimization);
diff --git a/src/main/java/net/minecraft/world/level/block/TrapDoorBlock.java b/src/main/java/net/minecraft/world/level/block/TrapDoorBlock.java
index b9e3c9cd709d42bced85436720ccc7b9c7e49552..24781846af3deeab313a6910d8a5b2ab97d41213 100644
--- a/src/main/java/net/minecraft/world/level/block/TrapDoorBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/TrapDoorBlock.java
@@ -1,6 +1,8 @@
 package net.minecraft.world.level.block;
 
 import javax.annotation.Nullable;
+
+import me.earthme.molia.MoliaConfig;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Direction;
 import net.minecraft.sounds.SoundSource;
@@ -124,29 +126,38 @@ public class TrapDoorBlock extends HorizontalDirectionalBlock implements SimpleW
                     flag1 = eventRedstone.getNewCurrent() > 0;
                 }
                 // CraftBukkit end
-                boolean open = (Boolean) state.getValue(TrapDoorBlock.OPEN) != flag1; // Paper - break redstone on trapdoors early
-                // Paper start - break redstone on trapdoors early
-                // note: this must run before any state for this block/its neighborus are written to the world
-                // we allow the redstone event to fire so that plugins can block
-                if (flag1 && open) { // if we are now powered and it caused the trap door to open
-                    // in this case, first check for the redstone on top first
-                    BlockPos abovePos = pos.above();
-                    BlockState above = world.getBlockState(abovePos);
-                    if (above.getBlock() instanceof RedStoneWireBlock) {
-                        world.setBlock(abovePos, Blocks.AIR.defaultBlockState(), Block.UPDATE_CLIENTS | Block.UPDATE_NEIGHBORS);
-                        Block.popResource(world, abovePos, new net.minecraft.world.item.ItemStack(net.minecraft.world.item.Items.REDSTONE));
-                        // now check that this didn't change our state
-                        if (world.getBlockState(pos) != state) {
-                            // our state was changed, so we cannot propagate this update
-                            return;
+                //Molia start - Add config for this fixes
+                if (MoliaConfig.breakRedstoneOnTripDoorEarly){
+                    boolean open = (Boolean) state.getValue(TrapDoorBlock.OPEN) != flag1; // Paper - break redstone on trapdoors early
+                    // Paper start - break redstone on trapdoors early
+                    // note: this must run before any state for this block/its neighborus are written to the world
+                    // we allow the redstone event to fire so that plugins can block
+                    if (flag1 && open) { // if we are now powered and it caused the trap door to open
+                        // in this case, first check for the redstone on top first
+                        BlockPos abovePos = pos.above();
+                        BlockState above = world.getBlockState(abovePos);
+                        if (above.getBlock() instanceof RedStoneWireBlock) {
+                            world.setBlock(abovePos, Blocks.AIR.defaultBlockState(), Block.UPDATE_CLIENTS | Block.UPDATE_NEIGHBORS);
+                            Block.popResource(world, abovePos, new net.minecraft.world.item.ItemStack(net.minecraft.world.item.Items.REDSTONE));
+                            // now check that this didn't change our state
+                            if (world.getBlockState(pos) != state) {
+                                // our state was changed, so we cannot propagate this update
+                                return;
+                            }
                         }
                     }
+                    // Paper end - break redstone on trapdoors early
+                    if (open) { // Paper - break redstone on trapdoors early
+                        state = (BlockState) state.setValue(TrapDoorBlock.OPEN, flag1);
+                        this.playSound((Player) null, world, pos, flag1);
+                    }
+                }else{
+                    if ((Boolean) state.getValue(TrapDoorBlock.OPEN) != flag1) {
+                        state = (BlockState) state.setValue(TrapDoorBlock.OPEN, flag1);
+                        this.playSound((Player) null, world, pos, flag1);
+                    }
                 }
-                // Paper end - break redstone on trapdoors early
-                if (open) { // Paper - break redstone on trapdoors early
-                    state = (BlockState) state.setValue(TrapDoorBlock.OPEN, flag1);
-                    this.playSound((Player) null, world, pos, flag1);
-                }
+                //Molia end
 
                 world.setBlock(pos, (BlockState) state.setValue(TrapDoorBlock.POWERED, flag1), 2);
                 if ((Boolean) state.getValue(TrapDoorBlock.WATERLOGGED)) {
