From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: M2ke4U <wangxyper@163.com>
Date: Sat, 24 Jun 2023 07:54:45 +0800
Subject: [PATCH] Fix config and leaves fakeplayer support


diff --git a/src/main/java/me/earthme/molia/MoliaConfig.java b/src/main/java/me/earthme/molia/MoliaConfig.java
index 3888fcb9410f30d1ed2efdb99b2f40a0dd5829c8..8dc09260c390fc988342e13f53d520a300a6170a 100644
--- a/src/main/java/me/earthme/molia/MoliaConfig.java
+++ b/src/main/java/me/earthme/molia/MoliaConfig.java
@@ -10,12 +10,21 @@ public class MoliaConfig {
     /**
      * Config values
      */
+    //Optimizations
     public static long checkStuckInWallInterval = 10;
     public static double entityWakeUpDurationRatioStandardDeviation = 0.2;
     public static int acquirePoiForStuckEntity = 200;
+
+    //Misc
     public static String serverModName = "Molia";
+
+    //Unsafe
     public static boolean disableFoliaPluginCheck = false;
+
+    //Gameplay
     public static boolean enableFakePlayer = true;
+
+    //Fake player
     public static boolean enableFakePlayerSkipSleep = true;
     public static int fakePlayerSizeLimit = 20;
     public static boolean alwaysSendFakePlayerData = true;
@@ -43,6 +52,12 @@ public class MoliaConfig {
         disableFoliaPluginCheck = getBoolean("unsafe.disable_folia_plugin_check",disableFoliaPluginCheck);
 
         enableFakePlayer = getBoolean("gameplay.enable_fake_player",enableFakePlayer);
+
+        enableFakePlayerSkipSleep = getBoolean("fakeplayer.enable_fake_player_skip_sleep",enableFakePlayerSkipSleep);
+        fakePlayerSizeLimit = getInt("fakeplayer.fake_player_size_limit",fakePlayerSizeLimit);
+        alwaysSendFakePlayerData = getBoolean("fakeplayer.always_send_fake_player_data",alwaysSendFakePlayerData);
+        canFakePlayerInteract = getBoolean("fakeplayer.can_fake_player_interact",canFakePlayerInteract);
+        fakeplayerResident = getBoolean("fakeplayer.fake_player_resident",fakeplayerResident);
         save();
     }
 
diff --git a/src/main/java/top/leavesmc/leaves/bot/ServerBot.java b/src/main/java/top/leavesmc/leaves/bot/ServerBot.java
index 5ef5f1c294e65674bc6b68ad4ca1b94e50b5990d..56a52ad145d7268a558f7431542956695dabec4d 100644
--- a/src/main/java/top/leavesmc/leaves/bot/ServerBot.java
+++ b/src/main/java/top/leavesmc/leaves/bot/ServerBot.java
@@ -139,9 +139,9 @@ public class ServerBot extends ServerPlayer {
             return;
         }
         if (skinName != null) {
-            Bukkit.getScheduler().runTaskAsynchronously(MINECRAFT_PLUGIN, () -> {
+            Bukkit.getRegionScheduler().execute(MINECRAFT_PLUGIN,loc,() -> {
                 String[] skin = MojangAPI.getSkin(skinName);
-                Bukkit.getScheduler().runTask(MINECRAFT_PLUGIN, () -> createBot(loc, name, name, skin, skinName, inv));
+                Bukkit.getRegionScheduler().execute(MINECRAFT_PLUGIN,loc, () -> createBot(loc, name, name, skin, skinName, inv));
             });
         } else {
             createBot(loc, name, name, null, null, inv);
