From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: M2ke4U <wangxyper@163.com>
Date: Sat, 24 Jun 2023 08:17:13 +0800
Subject: [PATCH] Fix concurrent issues in leaves patches


diff --git a/src/main/java/top/leavesmc/leaves/bot/MojangAPI.java b/src/main/java/top/leavesmc/leaves/bot/MojangAPI.java
index daaece30b2a3983f1cc9ee9a851e8f373974d5ec..9dcc44a8786c0442cd0812a8444d0c3f196f4b45 100644
--- a/src/main/java/top/leavesmc/leaves/bot/MojangAPI.java
+++ b/src/main/java/top/leavesmc/leaves/bot/MojangAPI.java
@@ -1,5 +1,6 @@
 package top.leavesmc.leaves.bot;
 
+import com.google.common.collect.Maps;
 import com.google.gson.JsonObject;
 import com.google.gson.JsonParser;
 
@@ -13,7 +14,7 @@ public class MojangAPI {
 
     private static final boolean CACHE_ENABLED = false;
 
-    private static final Map<String, String[]> CACHE = new HashMap<>();
+    private static final Map<String, String[]> CACHE = Maps.newConcurrentMap(); //Molia fix concurrent issue
 
     public static String[] getSkin(String name) {
         if (CACHE_ENABLED && CACHE.containsKey(name)) {
diff --git a/src/main/java/top/leavesmc/leaves/bot/ServerBot.java b/src/main/java/top/leavesmc/leaves/bot/ServerBot.java
index 56a52ad145d7268a558f7431542956695dabec4d..54d9440fa796751530e118e2a3c93787d2c54909 100644
--- a/src/main/java/top/leavesmc/leaves/bot/ServerBot.java
+++ b/src/main/java/top/leavesmc/leaves/bot/ServerBot.java
@@ -106,7 +106,7 @@ public class ServerBot extends ServerPlayer {
 
     private final BotInventoryContainer container;
 
-    private static final List<ServerBot> bots = new ArrayList<>();
+    private static final List<ServerBot> bots = Lists.newCopyOnWriteArrayList(); //Molia fix concurrent issue
     private static final Plugin MINECRAFT_PLUGIN = new MinecraftInternalPlugin();
 
     private ServerBot(MinecraftServer server, ServerLevel world, GameProfile profile) {
diff --git a/src/main/java/top/leavesmc/leaves/protocol/BBORProtocol.java b/src/main/java/top/leavesmc/leaves/protocol/BBORProtocol.java
index 4f2ac4079e78d0c6dc5e09a31f2fb1f15e83c544..f2dccfdb286f42cfbc6a385539aeb1abd8c40d0b 100644
--- a/src/main/java/top/leavesmc/leaves/protocol/BBORProtocol.java
+++ b/src/main/java/top/leavesmc/leaves/protocol/BBORProtocol.java
@@ -1,5 +1,6 @@
 package top.leavesmc.leaves.protocol;
 
+import com.google.common.collect.Maps;
 import io.netty.buffer.Unpooled;
 import me.earthme.molia.MoliaConfig;
 import net.minecraft.core.BlockPos;
@@ -37,7 +38,7 @@ public class BBORProtocol {
     // call
     public static final ResourceLocation SUBSCRIBE = id("subscribe");
     private static final Map<Integer, ServerPlayer> players = new ConcurrentHashMap<>();
-    private static final Map<Integer, Set<BBoundingBox>> playerBoundingBoxesCache = new HashMap<>();
+    private static final Map<Integer, Set<BBoundingBox>> playerBoundingBoxesCache = Maps.newConcurrentMap(); //Molia fix concurrent issue
     private static final Map<ResourceLocation, Map<BBoundingBox, Set<BBoundingBox>>> dimensionCache = new ConcurrentHashMap<>();
 
     private static final ServerLevel OVERWORLD = MinecraftServer.getServer().overworld();
