From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: M2ke4U <wangxyper@163.com>
Date: Sat, 24 Jun 2023 08:32:09 +0800
Subject: [PATCH] Compile error fixes


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 1f22962ae20911edef4adbc28a4aa71e357325f4..0a85dfa794c16546704d87d814a4a8eafbb8bcf6 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -29,6 +29,8 @@ import java.util.function.UnaryOperator;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import javax.annotation.Nullable;
+
+import me.earthme.molia.MoliaConfig;
 import net.minecraft.ChatFormatting;
 import net.minecraft.CrashReport;
 import net.minecraft.CrashReportCategory;
@@ -3675,10 +3677,10 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Tic
         } else {
             // Leaves start - leaves extra protocol
             try {
-                if (top.leavesmc.leaves.LeavesConfig.pcaSyncProtocol && ProtocolUtils.isNamespacePacket(packet, top.leavesmc.leaves.protocol.PcaSyncProtocol.PROTOCOL_ID)) {
+                if (MoliaConfig.enablePCASyncProtocol && ProtocolUtils.isNamespacePacket(packet, top.leavesmc.leaves.protocol.PcaSyncProtocol.PROTOCOL_ID)) {
                     top.leavesmc.leaves.protocol.PcaSyncProtocol.handlePacket(server, player, packet);
                 }
-                if (top.leavesmc.leaves.LeavesConfig.bborProtocol && packet.identifier.equals(top.leavesmc.leaves.protocol.BBORProtocol.SUBSCRIBE)) {
+                if (MoliaConfig.enableBBORProtocol && packet.identifier.equals(top.leavesmc.leaves.protocol.BBORProtocol.SUBSCRIBE)) {
                     top.leavesmc.leaves.protocol.BBORProtocol.onPlayerSubscribed(player);
                 }
             } catch (Exception ex) {
diff --git a/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index 59e9b446eb59d903ebe133f1c9a76bf41bbdc8dc..4fb423c6f30ac8fb432319425850bba907ef15c3 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -182,7 +182,7 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
                     blockEntity.setCooldown(world.spigotConfig.hopperTransfer); // Spigot
                     setChanged(world, pos, state);
                     // Leaves start - pca
-                    if (top.leavesmc.leaves.LeavesConfig.pcaSyncProtocol) {
+                    if (MoliaConfig.enablePCASyncProtocol) {
                         top.leavesmc.leaves.protocol.PcaSyncProtocol.syncBlockEntityToClient(blockEntity);
                     }
                     // Leaves end - pca
