From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Wed, 23 Aug 2023 15:25:19 +0800
Subject: [PATCH] Fix NPE in TpsBar


diff --git a/src/main/java/me/earthme/molia/functions/GlobalServerTpsBar.java b/src/main/java/me/earthme/molia/functions/GlobalServerTpsBar.java
index e93daefd27a831deda5faa5c596a8e3c587c03ec..1542561643cc10df7d217606fd10cb6f65263a1d 100644
--- a/src/main/java/me/earthme/molia/functions/GlobalServerTpsBar.java
+++ b/src/main/java/me/earthme/molia/functions/GlobalServerTpsBar.java
@@ -58,14 +58,25 @@ public class GlobalServerTpsBar {
     }
 
     private static void updateBarValues(){
+        final List<ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData>> regionsNeedRemove = new ArrayList<>();
         for (Map.Entry<ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData>, BossBar> entry : bossBars.entrySet()){
             final ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData> region = entry.getKey();
             final BossBar bossBar = entry.getValue();
 
-            final TickData.TickReportData reportData = region.getData().getRegionSchedulingHandle().getTickReport5s(System.nanoTime());
-            final TickData.SegmentData tpsData = reportData.tpsData().segmentAll();
-            final double mspt = reportData.timePerTickData().segmentAll().average() / 1.0E6;
-            updateTpsBar(tpsData.average(),mspt,bossBar);
+            if (region.getData().getRegionStats().getPlayerCount() != 0){ //Skip and remove the tpsbar if there are no players in the region
+                final TickData.TickReportData reportData = region.getData().getRegionSchedulingHandle().getTickReport5s(System.nanoTime());
+                final TickData.SegmentData tpsData = reportData.tpsData().segmentAll();
+                final double mspt = reportData.timePerTickData().segmentAll().average() / 1.0E6;
+
+                updateTpsBar(tpsData.average(),mspt,bossBar);
+                continue;
+            }
+
+            regionsNeedRemove.add(region);
+        }
+
+        for (ThreadedRegionizer.ThreadedRegion<TickRegions.TickRegionData, TickRegions.TickRegionSectionData> region : regionsNeedRemove){
+            bossBars.remove(region);
         }
     }
 
