From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Fri, 7 Jul 2023 10:20:38 +0800
Subject: [PATCH] Adjust some config key name and codes


diff --git a/src/main/java/me/earthme/molia/MoliaConfig.java b/src/main/java/me/earthme/molia/MoliaConfig.java
index b79af823b9dfe2e571efedf9e95b6eb0712f386c..6f5ac14f8534b3325da742382bdfa88a474a5a16 100644
--- a/src/main/java/me/earthme/molia/MoliaConfig.java
+++ b/src/main/java/me/earthme/molia/MoliaConfig.java
@@ -89,7 +89,7 @@ public class MoliaConfig {
         fakePlayerSizeLimit = getInt("fakeplayer.fake_player_size_limit",fakePlayerSizeLimit);
         alwaysSendFakePlayerData = getBoolean("fakeplayer.always_send_fake_player_data",alwaysSendFakePlayerData);
         canFakePlayerInteract = getBoolean("fakeplayer.can_fake_player_interact",canFakePlayerInteract);
-        fakeplayerResident = getBoolean("fakeplayer.fake_player_resident",fakeplayerResident);
+        fakePlayerResident = getBoolean("fakeplayer.fake_player_resident", fakePlayerResident);
         save();
     }
 
diff --git a/src/main/java/me/earthme/molia/thread/CraftSchedulerRunner.java b/src/main/java/me/earthme/molia/thread/CraftSchedulerRunner.java
index f730543bf166398a702ea7eaf7efaee8af346fd1..4f1ff25fb774e20c7e9c33368539072ebae6bc2e 100644
--- a/src/main/java/me/earthme/molia/thread/CraftSchedulerRunner.java
+++ b/src/main/java/me/earthme/molia/thread/CraftSchedulerRunner.java
@@ -4,16 +4,15 @@ import io.papermc.paper.util.TickThread;
 import net.minecraft.server.MinecraftServer;
 import org.bukkit.Bukkit;
 import org.bukkit.craftbukkit.CraftServer;
-import org.bukkit.craftbukkit.scheduler.CraftScheduler;
 
 import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.locks.LockSupport;
 
 /**
  * THIS CLASS ONLY USED FOR SUPPORT BUKKIT SCHEDULER
+ * WARNING:USE THIS WOULD MAKE BUKKIT PLUGINS CAN CRASH YOUR SERVER
  */
-@Deprecated
+@Deprecated(forRemoval = true)
 public class CraftSchedulerRunner extends TickThread {
     public static final CraftSchedulerRunner RUNNER_INSTANCE;
 
@@ -36,6 +35,7 @@ public class CraftSchedulerRunner extends TickThread {
         return Thread.currentThread() instanceof CraftSchedulerRunner;
     }
 
+    @Deprecated
     public int getTickCount() {
         return this.tickCount;
     }
diff --git a/src/main/java/top/leavesmc/leaves/bot/ServerBot.java b/src/main/java/top/leavesmc/leaves/bot/ServerBot.java
index 54d9440fa796751530e118e2a3c93787d2c54909..92b6fe3155fb57dc4e955b88361f5fafefa23b3b 100644
--- a/src/main/java/top/leavesmc/leaves/bot/ServerBot.java
+++ b/src/main/java/top/leavesmc/leaves/bot/ServerBot.java
@@ -710,7 +710,7 @@ public class ServerBot extends ServerPlayer {
     }
 
     public static void saveOrRemoveAllBot() {
-        if (me.earthme.molia.MoliaConfig.enableFakePlayer && me.earthme.molia.MoliaConfig.fakeplayerResident) {
+        if (me.earthme.molia.MoliaConfig.enableFakePlayer && me.earthme.molia.MoliaConfig.fakePlayerResident) {
             JsonObject fakePlayerList = new JsonObject();
             bots.forEach(bot -> fakePlayerList.add(bot.realName, BotUtil.saveBot(bot)));
             File file = MinecraftServer.getServer().getWorldPath(LevelResource.ROOT).resolve("fake_player.leaves.json").toFile();
@@ -732,7 +732,7 @@ public class ServerBot extends ServerPlayer {
     }
 
     public static void loadAllBot() {
-        if (me.earthme.molia.MoliaConfig.enableFakePlayer && me.earthme.molia.MoliaConfig.fakeplayerResident) {
+        if (me.earthme.molia.MoliaConfig.enableFakePlayer && me.earthme.molia.MoliaConfig.fakePlayerResident) {
             JsonObject fakePlayerList = new JsonObject();
             File file = MinecraftServer.getServer().getWorldPath(LevelResource.ROOT).resolve("fake_player.leaves.json").toFile();
             if (!file.isFile()) {
