From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Fri, 7 Jul 2023 10:31:51 +0800
Subject: [PATCH] Correct a config key name in dab


diff --git a/src/main/java/me/earthme/molia/MoliaConfig.java b/src/main/java/me/earthme/molia/MoliaConfig.java
index 6f5ac14f8534b3325da742382bdfa88a474a5a16..806a9c820d059b8879f924a7644801bdcb7ea647 100644
--- a/src/main/java/me/earthme/molia/MoliaConfig.java
+++ b/src/main/java/me/earthme/molia/MoliaConfig.java
@@ -70,7 +70,7 @@ public class MoliaConfig {
         dearEnabled = getBoolean("optimizations.dab.enabled",true);
         startDistanceSquared = getInt("optimizations.dab.start_distance",12) ^ 2;
         maximumActivationPrio = getInt("optimizations.dab.max_tick_freq",20);
-        activationDistanceMod = getInt("optimizations.activation_dist_mod",8);
+        activationDistanceMod = getInt("optimizations.dab.activation_dist_mod",8);
 
         prepareForDAB();
 
