From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Wed, 30 Aug 2023 11:12:26 +0800
Subject: [PATCH] Stop bossbar before call stopServer in the parent class


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 1c7402f2db03943431b29d7413ee4dd9b1e6163f..f7fa8cbe64c5cf8bfdc36e80325179cf1765b932 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -821,8 +821,8 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 
     @Override
     public void stopServer() {
-        super.stopServer();
         GlobalServerTpsBar.onServerStop(); //Molia
+        super.stopServer();
         //Util.shutdownExecutors(); // Paper - moved into super
         SkullBlockEntity.clear();
     }
