From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Sat, 9 Sep 2023 21:58:26 +0800
Subject: [PATCH] Fix tpsbar visible exclude issue


diff --git a/src/main/java/me/earthme/molia/functions/GlobalServerTpsBar.java b/src/main/java/me/earthme/molia/functions/GlobalServerTpsBar.java
index 71a264fa7c16946ffb8dbac0273c4a05c8827700..5c852426ef251a77396a9921ac2c2e2b3ff91ea0 100644
--- a/src/main/java/me/earthme/molia/functions/GlobalServerTpsBar.java
+++ b/src/main/java/me/earthme/molia/functions/GlobalServerTpsBar.java
@@ -105,7 +105,12 @@ public class GlobalServerTpsBar {
 
     private static void updateBarVisibilityForPlayers(){
         for (Player player : Bukkit.getOnlinePlayers()){
-            if (player instanceof Bot || visibleExclude.contains(player)){
+            if (player instanceof Bot){
+                continue;
+            }
+
+            if (visibleExclude.contains(player)){
+                addedPlayers.remove(player);
                 continue;
             }
 
