From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Wed, 19 Jul 2023 22:03:16 +0800
Subject: [PATCH] Readd config for Rework patch Fix can not move entity off
 main


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 2310f588943e8885612a5e807ebcb0330309c072..f5ef2b0deaea62a2333ca0690d09af0317243f25 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -1129,18 +1129,22 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
                     }
                 }
 
-                var targetX = this.getX() + vec3d1.x;
-                var targetZ = this.getZ() + vec3d1.z;
-
-                if (!TickThread.isTickThreadFor(((ServerLevel) this.level()),targetX,targetZ)) {
-                    this.teleportAsync((ServerLevel) this.level(), this.position().add(vec3d1),
-                            this.getYRot(),
-                            this.getXRot(),
-                            null,
-                            PlayerTeleportEvent.TeleportCause.UNKNOWN,
-                            Entity.TELEPORT_FLAG_LOAD_CHUNK | Entity.TELEPORT_FLAG_TELEPORT_PASSENGERS,
-                            null
-                    );
+                if (MoliaConfig.fixMoveEntityOffMain){
+                    var targetX = this.getX() + vec3d1.x;
+                    var targetZ = this.getZ() + vec3d1.z;
+
+                    if (!TickThread.isTickThreadFor(((ServerLevel) this.level()),targetX,targetZ)) {
+                        this.teleportAsync((ServerLevel) this.level(), this.position().add(vec3d1),
+                                this.getYRot(),
+                                this.getXRot(),
+                                null,
+                                PlayerTeleportEvent.TeleportCause.UNKNOWN,
+                                Entity.TELEPORT_FLAG_LOAD_CHUNK | Entity.TELEPORT_FLAG_TELEPORT_PASSENGERS,
+                                null
+                        );
+                    }else{
+                        this.setPos(this.getX() + vec3d1.x, this.getY() + vec3d1.y, this.getZ() + vec3d1.z);
+                    }
                 }else{
                     this.setPos(this.getX() + vec3d1.x, this.getY() + vec3d1.y, this.getZ() + vec3d1.z);
                 }
