From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: wangxyper <wangxyper@163.com>
Date: Wed, 30 Aug 2023 11:12:26 +0800
Subject: [PATCH] Stop bossbar before call stopServer in the parent class


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index cff91fee161809f1a0a043fca66d4a5c888cb6e5..8afd70ee2682531f8153415ca0185116f557b3a9 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -814,8 +814,8 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 
     @Override
     public void stopServer() {
-        super.stopServer();
         GlobalServerTpsBar.onServerStop(); //Molia
+        super.stopServer();
         //Util.shutdownExecutors(); // Paper - moved into super
         SkullBlockEntity.clear();
     }
