From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Sun, 16 Jul 2023 20:42:16 +0800
Subject: [PATCH] Add config keys for fixes


diff --git a/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java b/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java
index 69a9dc0aa53d4bf1ae841db9354b377dafedab8d..f6fc37663b95c4384295f8b269b8102670cba7fd 100644
--- a/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java
+++ b/src/main/java/io/papermc/paper/chunk/system/RegionizedPlayerChunkLoader.java
@@ -15,6 +15,7 @@ import it.unimi.dsi.fastutil.longs.LongArrayList;
 import it.unimi.dsi.fastutil.longs.LongComparator;
 import it.unimi.dsi.fastutil.longs.LongHeapPriorityQueue;
 import it.unimi.dsi.fastutil.longs.LongOpenHashSet;
+import me.earthme.molia.MoliaConfig;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.game.ClientboundSetChunkCacheCenterPacket;
 import net.minecraft.network.protocol.game.ClientboundSetChunkCacheRadiusPacket;
@@ -481,7 +482,8 @@ public class RegionizedPlayerChunkLoader {
 
         private void sendChunk(final int chunkX, final int chunkZ) {
             //Molia start
-            if (this.sentChunks.contains(CoordinateUtils.getChunkKey(chunkX, chunkZ))){
+            //Molia - add config for this
+            if (this.sentChunks.contains(CoordinateUtils.getChunkKey(chunkX, chunkZ)) && MoliaConfig.fixDoubleChunkSend){
                 return;
             }
             //Molia end
diff --git a/src/main/java/me/earthme/molia/MoliaConfig.java b/src/main/java/me/earthme/molia/MoliaConfig.java
index 95fe40b7003a92d83787ed92e829c9d28d5dd54c..37b8b08c541fc75a047a8a18a7c007caa6881d5b 100644
--- a/src/main/java/me/earthme/molia/MoliaConfig.java
+++ b/src/main/java/me/earthme/molia/MoliaConfig.java
@@ -21,6 +21,8 @@ public class MoliaConfig {
 
     //Misc
     public static String serverModName = "Molia";
+    public static boolean fixMoveEntityOffMain = true;
+    public static boolean fixDoubleChunkSend = true;
 
     //Unsafe
     public static boolean disableFoliaPluginCheck = false;
@@ -79,6 +81,8 @@ public class MoliaConfig {
         prepareForDAB();
 
         serverModName = getString("misc.server_mod_name","Molia");
+        fixDoubleChunkSend = getBoolean("misc.fix_double_chunk_send",fixDoubleChunkSend);
+        fixMoveEntityOffMain = getBoolean("misc.fix_move_entity_off_main",fixMoveEntityOffMain);
 
         disableFoliaPluginCheck = getBoolean("unsafe.disable_folia_plugin_check",disableFoliaPluginCheck);
 
