From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Sun, 6 Aug 2023 17:06:10 +0800
Subject: [PATCH] Add config for sand duping


diff --git a/src/main/java/me/earthme/molia/MoliaConfig.java b/src/main/java/me/earthme/molia/MoliaConfig.java
index 2b4d48e41c5df971bad1492f3d7c7b2b14a7b0f0..24b9a555f4de81373a49e0a436f1701cd6d4ad44 100644
--- a/src/main/java/me/earthme/molia/MoliaConfig.java
+++ b/src/main/java/me/earthme/molia/MoliaConfig.java
@@ -32,6 +32,7 @@ public class MoliaConfig {
     public static String pcaSyncPlayerEntity = "NOBODY";
     public static boolean enableXeroMapProtocol = false;
     public static int xeroServerId = 1;
+    public static boolean disableSandDupingFix = false;
 
     //Fake player
     public static boolean enableFakePlayerSkipSleep = true;
@@ -79,6 +80,7 @@ public class MoliaConfig {
         pcaSyncPlayerEntity = getString("gameplay.pca_sync_player_entity",pcaSyncPlayerEntity);
         enableXeroMapProtocol = getBoolean("gameplay.enable_xero_map_protocol",enableXeroMapProtocol);
         xeroServerId = getInt("gameplay.xero_server_id",1);
+        disableSandDupingFix = getBoolean("gameplay.disable_sand_duping_fixes",disableSandDupingFix);
 
         enableFakePlayerSkipSleep = getBoolean("fakeplayer.enable_fake_player_skip_sleep",enableFakePlayerSkipSleep);
         fakePlayerSizeLimit = getInt("fakeplayer.fake_player_size_limit",fakePlayerSizeLimit);
diff --git a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
index 146254dc045e529a0fc4622a91c7b6d022a6b42d..739791e22296e76c74dc243cff5004db503fafd4 100644
--- a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -133,11 +133,13 @@ public class FallingBlockEntity extends Entity {
     @Override
     public void tick() {
         //Molia start - we don't need this
-        /*// Paper start - fix sand duping
-        if (this.isRemoved()) {
+        //Molia start - add config for this
+        // Paper start - fix sand duping
+        if (this.isRemoved() && me.earthme.molia.MoliaConfig.disableSandDupingFix) {
             return;
         }
-        // Paper end - fix sand duping*/
+        // Paper end - fix sand duping
+        //Molia end
         //Molia end
         if (this.blockState.isAir()) {
             this.discard();
@@ -152,11 +154,13 @@ public class FallingBlockEntity extends Entity {
             this.move(MoverType.SELF, this.getDeltaMovement());
 
             //Molia start - we don't need this
-            /*// Paper start - fix sand duping
-            if (this.isRemoved()) {
+            //Molia start - add config
+            // Paper start - fix sand duping
+            if (this.isRemoved() && me.earthme.molia.MoliaConfig.disableSandDupingFix) {
                 return;
             }
-            // Paper end - fix sand duping*/
+            //Molia end
+            // Paper end - fix sand duping
             //Molia end
 
             // Paper start - Configurable EntityFallingBlock height nerf
