From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Wed, 9 Aug 2023 09:23:41 +0800
Subject: [PATCH] Fix NPE caused by getWorldData off region thread in leaves's
 fakeplayer


diff --git a/src/main/java/top/leavesmc/leaves/bot/ServerBot.java b/src/main/java/top/leavesmc/leaves/bot/ServerBot.java
index 92b6fe3155fb57dc4e955b88361f5fafefa23b3b..50ec33b6b030ca7e2dbfbbfead7690ad6eeb1947 100644
--- a/src/main/java/top/leavesmc/leaves/bot/ServerBot.java
+++ b/src/main/java/top/leavesmc/leaves/bot/ServerBot.java
@@ -144,7 +144,9 @@ public class ServerBot extends ServerPlayer {
                 Bukkit.getRegionScheduler().execute(MINECRAFT_PLUGIN,loc, () -> createBot(loc, name, name, skin, skinName, inv));
             });
         } else {
-            createBot(loc, name, name, null, null, inv);
+            //Move to region scheduler to ensure thread safe
+            //createBot(loc, name, name, null, null, inv);
+            Bukkit.getRegionScheduler().execute(MINECRAFT_PLUGIN,loc,()-> createBot(loc, name, name, null, null, inv));
         }
     }
 
