From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Wed, 9 Aug 2023 09:33:59 +0800
Subject: [PATCH] Make fakeplayer join or left message more like normal player


diff --git a/src/main/java/top/leavesmc/leaves/bot/ServerBot.java b/src/main/java/top/leavesmc/leaves/bot/ServerBot.java
index 50ec33b6b030ca7e2dbfbbfead7690ad6eeb1947..8f5cfd0d20e00614e4c96f64e3f8cd94a12e6a28 100644
--- a/src/main/java/top/leavesmc/leaves/bot/ServerBot.java
+++ b/src/main/java/top/leavesmc/leaves/bot/ServerBot.java
@@ -7,10 +7,14 @@ import com.google.gson.JsonObject;
 import com.mojang.authlib.GameProfile;
 import com.mojang.authlib.properties.Property;
 import com.mojang.datafixers.util.Pair;
+import io.papermc.paper.adventure.PaperAdventure;
+import io.papermc.paper.configuration.GlobalConfiguration;
+import net.kyori.adventure.text.format.NamedTextColor;
 import net.minecraft.Util;
 import net.minecraft.nbt.ListTag;
 import net.minecraft.network.Connection;
 import net.minecraft.network.PacketSendListener;
+import net.minecraft.network.chat.Component;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.PacketFlow;
 import net.minecraft.network.protocol.game.ClientboundAddPlayerPacket;
@@ -177,7 +181,13 @@ public class ServerBot extends ServerPlayer {
             bot.isRealPlayer = true;
 
             if (event.getJoinMessage() != null) {
-                Bukkit.broadcastMessage(event.getJoinMessage());
+                //Molia start - Make fake player more like normal player if the join message is not modified
+                if (event.isJoinMessageModified()){
+                    Bukkit.broadcastMessage(event.getJoinMessage());
+                }else{
+                    Bukkit.broadcast((net.kyori.adventure.text.Component) Component.translatable("multiplayer.player.joined", bot.getDisplayName()));
+                }
+                //Molia end
             }
 
             if (inv != null) {
@@ -263,7 +273,7 @@ public class ServerBot extends ServerPlayer {
             remove(RemovalReason.KILLED);
             this.setDead();
             this.removeTab();
-            Bukkit.broadcastMessage(ChatColor.YELLOW + this.getName().getString() + " leaved the game"); // TODO i18n
+            Bukkit.broadcast((net.kyori.adventure.text.Component) Component.translatable("multiplayer.player.left", NamedTextColor.YELLOW, GlobalConfiguration.get().messages.useDisplayNameInQuitMessage ? this.getBukkitEntity().displayName() : PaperAdventure.asAdventure(this.getDisplayName()))); // TODO i18n //Molia - Make fake player more like normal player if the join message is not modified
         }
     }
 
