From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Wed, 9 Aug 2023 21:50:21 +0800
Subject: [PATCH] Fix my stupid logic


diff --git a/src/main/java/top/leavesmc/leaves/bot/ServerBot.java b/src/main/java/top/leavesmc/leaves/bot/ServerBot.java
index 50ec33b6b030ca7e2dbfbbfead7690ad6eeb1947..503140dc85f4b9347cbcb8aa7c8d37ce76ca05a8 100644
--- a/src/main/java/top/leavesmc/leaves/bot/ServerBot.java
+++ b/src/main/java/top/leavesmc/leaves/bot/ServerBot.java
@@ -139,7 +139,7 @@ public class ServerBot extends ServerPlayer {
             return;
         }
         if (skinName != null) {
-            Bukkit.getRegionScheduler().execute(MINECRAFT_PLUGIN,loc,() -> {
+            Bukkit.getAsyncScheduler().runNow(MINECRAFT_PLUGIN,(asyncTask) -> {
                 String[] skin = MojangAPI.getSkin(skinName);
                 Bukkit.getRegionScheduler().execute(MINECRAFT_PLUGIN,loc, () -> createBot(loc, name, name, skin, skinName, inv));
             });
