From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Sun, 23 Jul 2023 08:58:32 +0800
Subject: [PATCH] Adjust config init call


diff --git a/src/main/java/me/earthme/molia/MoliaConfig.java b/src/main/java/me/earthme/molia/MoliaConfig.java
index 37b8b08c541fc75a047a8a18a7c007caa6881d5b..8b903569e38dcae1951c6243a0f4b8e4f609bf92 100644
--- a/src/main/java/me/earthme/molia/MoliaConfig.java
+++ b/src/main/java/me/earthme/molia/MoliaConfig.java
@@ -57,16 +57,7 @@ public class MoliaConfig {
         }
     }
 
-    public static void init(){
-        try {
-            configEntry.load(CONFIG_FILE);
-        }catch (IOException ignored){
-        } catch (InvalidConfigurationException e) {
-            e.printStackTrace();
-        }
-        configEntry.options().copyDefaults(true);
-
-        //Init the config values
+    public static void loadAllValues(){
         checkStuckInWallInterval = getLong("optimizations.check_stuck_in_wall_interval",checkStuckInWallInterval);
         entityWakeUpDurationRatioStandardDeviation = getDouble("optimizations.entity_wake_up_duration_ratio_standard_deviation",entityWakeUpDurationRatioStandardDeviation);
         acquirePoiForStuckEntity = getInt("optimizations.acquire_poi_for_stuck_entity",acquirePoiForStuckEntity);
@@ -98,6 +89,18 @@ public class MoliaConfig {
         alwaysSendFakePlayerData = getBoolean("fakeplayer.always_send_fake_player_data",alwaysSendFakePlayerData);
         canFakePlayerInteract = getBoolean("fakeplayer.can_fake_player_interact",canFakePlayerInteract);
         fakePlayerResident = getBoolean("fakeplayer.fake_player_resident", fakePlayerResident);
+    }
+
+    public static void init(){
+        try {
+            configEntry.load(CONFIG_FILE);
+        }catch (IOException ignored){
+        } catch (InvalidConfigurationException e) {
+            e.printStackTrace();
+        }
+        configEntry.options().copyDefaults(true);
+
+        loadAllValues();
         save();
     }
 
