From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Wed, 9 Aug 2023 21:50:21 +0800
Subject: [PATCH] Fix my stupid logic


diff --git a/src/main/java/top/leavesmc/leaves/bot/ServerBot.java b/src/main/java/top/leavesmc/leaves/bot/ServerBot.java
index 8f5cfd0d20e00614e4c96f64e3f8cd94a12e6a28..48b241b4f7f08f4fb6e01ed6a606a1f2bf5466b9 100644
--- a/src/main/java/top/leavesmc/leaves/bot/ServerBot.java
+++ b/src/main/java/top/leavesmc/leaves/bot/ServerBot.java
@@ -143,7 +143,7 @@ public class ServerBot extends ServerPlayer {
             return;
         }
         if (skinName != null) {
-            Bukkit.getRegionScheduler().execute(MINECRAFT_PLUGIN,loc,() -> {
+            Bukkit.getAsyncScheduler().runNow(MINECRAFT_PLUGIN,(asyncTask) -> {
                 String[] skin = MojangAPI.getSkin(skinName);
                 Bukkit.getRegionScheduler().execute(MINECRAFT_PLUGIN,loc, () -> createBot(loc, name, name, skin, skinName, inv));
             });
